htlcswitch/payment_result.go

```diff
@@ -0,0 +1,268 @@
+package htlcswitch
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+
+	// paymentStoreBucketKey is used for the root level bucket that
+	// stores the payment result for each payment ID.
+	paymentStoreBucketKey = []byte("payment-result-store-bucket")
+
+	// ErrPaymentIDNotFound is an error returned if the given paymentID is
+	// not found.
+	ErrPaymentIDNotFound = errors.New("paymentID not found")
+
+	// ErrPaymentIDAlreadyExists is returned if we try to write a pending
+	// payment whose paymentID already exists.
+	ErrPaymentIDAlreadyExists = errors.New("paymentID already exists")
+)
+
+// PaymentResult wraps a result received from the network after a payment
+// attempt was made.
+type PaymentResult struct {
+	// Preimage is set by the switch in case a sent HTLC was settled.
+	Preimage [32]byte
+
+	// Error is non-nil in case a HTLC send failed, and the HTLC is now
+	// irrevocably cancelled.
+	Error *ForwardingError
+}
+
+// serializePaymentResult serializes the PaymentResult.
+func serializePaymentResult(w io.Writer, p *PaymentResult) error {
+	if err := channeldb.WriteElements(w,
+		p.Preimage[:],
+	); err != nil {
+		return err
+	}
+
+	// If there's no error, write 'false' and return.
+	if p.Error == nil {
+		return channeldb.WriteElements(w, false)
+	}
+
+	// Otherwise, write 'true' and the error fields.
+	msg := []byte(p.Error.ExtraMsg)
+	err := channeldb.WriteElements(w, true,
```

Comment:
Saving the human readable error `ExtraMsg` to the db is not ideal. There are only a few cases for `ExtraMsg`, those can probably be saved in a structured way.

Reply:
Upon further inspection, there may be a few more. But still doubt whether we should persist data just to get a nice error message. Maybe just logging when the error occurs is enough.

Reply:
Yeah, looks like they are only used locally for passing the case of error to the RPC. Would it be useful to define a field for `local error codes` that can be used to communicate local failures?

Reply:
I think that local internal errors (things that should never happen) don't need to be exposed over rpc. Logging and returning a generic internal error should be enough?